addons/mail/models/ir_websocket.py

```diff
@@ -31,35 +36,65 @@ def _update_mail_presence(self, inactivity_period):
             self.env.user if partner else guest, inactivity_period
         )
 
-    def _filter_accessible_presences(self, partners, guests):
-        """Filter presences that are accessible to current user."""
-        if self.env.user and self.env.user._is_internal():
-            return partners, guests
-        return self.env["res.partner"], self.env["mail.guest"]
-
     def _prepare_subscribe_data(self, channels, last):
         data = super()._prepare_subscribe_data(channels, last)
-        str_presence_channels = {
-            c for c in channels if isinstance(c, str) and c.startswith("odoo-presence-")
-        }
-        simplified_presence_channels = [
-            tuple(c.replace("odoo-presence-", "").split("_")) for c in str_presence_channels
-        ]
-        for channel in str_presence_channels:
+        partner_id_to_token = {}
+        guest_id_to_token = {}
+
+        for channel in channels:
+            if not isinstance(channel, str) or not channel.startswith("odoo-presence-"):
+                continue
             data["channels"].discard(channel)
-        partner_ids = [int(p[1]) for p in simplified_presence_channels if p[0] == "res.partner"]
-        partners = (
+            if not (
+                match := re.match(
+                    r"^odoo-presence-(res\.partner|mail\.guest)_(\d+)(?:-(.+))?$", channel
+                )
+            ):
+                _logger.warning("Malformed presence channel: %s", channel)
+                continue
+            model, record_id, token = match.groups()
+            record_id = int(record_id)
+            if model == "res.partner":
+                partner_id_to_token[record_id] = token or ""
+            elif model == "mail.guest":
+                guest_id_to_token[record_id] = token or ""
```

Comment:
it feels like it could be a single line: `model_ids_to_token[model][int(record_id)] = token`

Reply:
I thought it was odd to have a generic dict while we only support two models, but why not :+1: